src/main/java/fi/metatavu/metamind/server/rest/StoriesApiImpl.java

```diff
@@ -461,57 +456,71 @@ public Response findVariable(UUID storyId, UUID variableId) {
 
   @Override
   public Response listIntents(UUID storyId) {
-    // TODO: Permission check
-    
     fi.metatavu.metamind.persistence.models.Story story = storyController.findStoryById(storyId);
     if (story == null) {
       return createBadRequest(String.format("Story %s not found", storyId)); 
     }
     
+    List<fi.metatavu.metamind.persistence.models.Intent> intents = storyController.listIntentsByStory(story);
+    if (intents.isEmpty()) {
+      return createOk(new ArrayList<fi.metatavu.metamind.persistence.models.Intent>());
+    }
+    
     return createOk(storyController.listIntentsByStory(story).stream()
       .map(intent -> intentTranslator.translateIntent(intent, trainingMaterialController.listTrainingMaterialByIntent(intent)))
       .collect(Collectors.toList()));
   }
 
   @Override
   public Response listKnots(UUID storyId) {
-    // TODO: Permission check
-    
     fi.metatavu.metamind.persistence.models.Story story = storyController.findStoryById(storyId);
     if (story == null) {
       return createBadRequest(String.format("Story %s not found", storyId)); 
     }
     
+    List<fi.metatavu.metamind.persistence.models.Knot> knots = storyController.listKnotsByStory(story);
+    if (knots.isEmpty()) {
+      return createOk(new ArrayList<fi.metatavu.metamind.persistence.models.Knot>());
+    }
+
     return createOk(storyController.listKnotsByStory(story).stream()
       .map(knotTranslator::translateKnot)
       .collect(Collectors.toList()));
   }
 
   @Override
   public Response listStories() {
-    // TODO: Permission check
+    UUID loggedUserId = getLoggerUserId();
```

Comment:
After the getPermittedStories method is fixed (see comment above), you can simplify this code a lot.

e.g.
     return createOk(authenticationController.getPermittedStories(loggedUserId, storyController.listStories()).stream()